### ISSUE TITLE
Reviewer - Negative Review Count after Undo

### ISSUE THREAD

**Issue by infinyte7** [ankidroid/Anki-Android#6587]
###### Reproduction Steps

1.  Create new deck
2.  Add two-three cards
3.  Review card
4. Undo Review

###### Expected Result
It should reset review to initial state.

###### Actual Result
It reset but in minus.
![desktop](https://user-images.githubusercontent.com/12841290/86093899-4c145e80-bae2-11ea-8072-175417f24a41.PNG)

###### Debug info
AnkiDroid 2.12alpha18

###### Research
*Enter an [x] character to confirm the points below:*

- [ x ] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
- [ x ] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [ x ] I have searched for similar existing issues here and on the user forum
- [ x ] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

**Comment by david-allison-1** [comment]
Thanks for the report!

Yeah, that's a bug

**Comment by david-allison-1** [comment]
Hmm... I have an inkling this may be a bug in my test, but:

```
34892482d021417cdfd5f9c8bcded1d9715cb237 is the first bad commit
commit 34892482d021417cdfd5f9c8bcded1d9715cb237
Author: Arthur Milchior <[redacted]>
Date:   Tue Apr 14 13:21:41 2020 +0200

    reset -> planifyReset in col.reset

    Col.reset is called 5 times. In no case the result of the scheduler is
    directly needed. So it's an acceptable change

 AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
bisect run success

C:\GitHub\Anki-Android-David>

```

<details><summary>6587.patch</summary>

```
Index: AnkiDroid/src/test/java/com/ichi2/anki/RobolectricTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- AnkiDroid/src/test/java/com/ichi2/anki/RobolectricTest.java	(revision 34d1808e1d14538878725234e90d8dd831265c29)
+++ AnkiDroid/src/test/java/com/ichi2/anki/RobolectricTest.java	(revision 9c85b12ddba862ffec56142e31d23ec3dd1a06b5)
@@ -23,6 +23,7 @@
 import com.afollestad.materialdialogs.MaterialDialog;
 import com.ichi2.anki.dialogs.DialogHandler;
 import com.ichi2.anki.exception.ConfirmModSchemaException;
+import com.ichi2.async.CollectionTask;
 import com.ichi2.compat.customtabs.CustomTabActivityHelper;
 import com.ichi2.libanki.Collection;
 import com.ichi2.libanki.DB;
@@ -239,4 +240,30 @@
 
         return (SchedV2) sched;
     }
+
+
+    protected synchronized void waitForTask(int taskType, int timeoutMs) throws InterruptedException {
+        boolean[] completed = new boolean[] { false };
+        CollectionTask.launchCollectionTask(taskType, new CollectionTask.TaskListener() {
+            @Override
+            public void onPreExecute() {
+
+            }
+
+
+            @Override
+            public void onPostExecute(CollectionTask.TaskData result) {
+                completed[0] = true;
+                synchronized (RobolectricTest.this) {
+                    RobolectricTest.this.notify();
+                }
+            }
+        });
+
+        wait(timeoutMs);
+
+        if (!completed[0]) {
+            throw new IllegalStateException(String.format("Task %d didn't finish in %d ms", taskType, timeoutMs));
+        }
+    }
 }
Index: AnkiDroid/src/test/java/com/ichi2/libanki/sched/AbstractSchedTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- AnkiDroid/src/test/java/com/ichi2/libanki/sched/AbstractSchedTest.java	(revision 9c85b12ddba862ffec56142e31d23ec3dd1a06b5)
+++ AnkiDroid/src/test/java/com/ichi2/libanki/sched/AbstractSchedTest.java	(revision 9c85b12ddba862ffec56142e31d23ec3dd1a06b5)
@@ -0,0 +1,86 @@
+/*
+ Copyright (c) 2020 David Allison <davidallisongithub@gmail.com>
+
+ This program is free software; you can redistribute it and/or modify it under
+ the terms of the GNU General Public License as published by the Free Software
+ Foundation; either version 3 of the License, or (at your option) any later
+ version.
+
+ This program is distributed in the hope that it will be useful, but WITHOUT ANY
+ WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+ PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License along with
+ this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.ichi2.libanki.sched;
+
+import com.ichi2.anki.RobolectricTest;
+import com.ichi2.anki.exception.ConfirmModSchemaException;
+import com.ichi2.async.CollectionTask;
+import com.ichi2.libanki.Card;
+import com.ichi2.libanki.Consts;
+
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.ParameterizedRobolectricTestRunner;
+import org.robolectric.ParameterizedRobolectricTestRunner.Parameter;
+import org.robolectric.ParameterizedRobolectricTestRunner.Parameters;
+
+import java.util.Arrays;
+
+import static com.ichi2.anki.AbstractFlashcardViewer.EASE_3;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.not;
+
+// Note: These tests can't be run individually but can from the class-level
+// gradlew AnkiDroid:testDebug --tests "com.ichi2.libanki.sched.AbstractSchedTest.*"
+@RunWith(ParameterizedRobolectricTestRunner.class)
+public class AbstractSchedTest extends RobolectricTest {
+
+    @Parameter
+    public int schedVersion;
+
+    @Parameters(name = "SchedV{0}")
+    public static java.util.Collection<Object[]> initParameters() {
+        // This does one run with schedVersion injected as 1, and one run as 2
+        return Arrays.asList(new Object[][] { { 1 }, { 2 } });
+    }
+
+    @Before
+    @Override
+    public void setUp() {
+        super.setUp();
+        try {
+            getCol().changeSchedulerVer(schedVersion);
+        } catch (ConfirmModSchemaException e) {
+            throw new RuntimeException("Could not change schema", e);
+        }
+    }
+
+    @Test
+    public void testUndoResetsCardCountsToCorrectValue() throws InterruptedException {
+        // #6587
+        addNoteUsingBasicModel("Hello", "World");
+
+        AbstractSched sched = getCol().getSched();
+        sched.reset();
+
+        Card cardBeforeUndo = sched.getCard();
+        int[] countsBeforeUndo = sched.counts();
+        // Not shown in the UI, but there is a state where the card has been removed from the queue, but not answered
+        // where the counts are decremented.
+        assertThat(countsBeforeUndo, is(new int[] { 0, 0, 0 }));
+
+        sched.answerCard(cardBeforeUndo, EASE_3);
+
+        waitForTask(CollectionTask.TASK_TYPE_UNDO, 5000);
+
+        int[] countsAfterUndo = sched.counts();
+
+        assertThat("Counts after an undo should be the same as before an undo", countsAfterUndo, is(countsBeforeUndo));
+    }
+}
```
</details>

**Comment by Arthur-Milchior** [comment]
I'll take a look later. Probably me. I don't remember what was and wasn't merged in 2.1.18alpha, so hard to diagnose on my phonr
